cni/pkg/nodeagent/ztunnelserver.go

```diff
@@ -0,0 +1,480 @@
+			// something first, we expect to get an os.ErrDeadlineExceeded error
+			// here if the connection is still alive.
+			// note that unlike tcp connections, reading is a good enough test here.
+			_, err := conn.readMessage(time.Second / 100)
+			switch {
+			case !errors.Is(err, os.ErrDeadlineExceeded):
+				log.Debugf("ztunnel keepalive failed: %v", err)
+				if errors.Is(err, io.EOF) {
+					log.Debug("ztunnel EOF")
+					return nil
+				}
+				return err
+			case err == nil:
+				log.Warn("ztunnel protocol error, unexpected message")
+				return fmt.Errorf("ztunnel protocol error, unexpected message")
+			default:
+				// we get here if error is deadline exceeded, which means ztunnel is alive.
+			}
+
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (z *ztunnelServer) PodDeleted(ctx context.Context, uid string) error {
+	r := &zdsapi.WorkloadRequest{
+		Payload: &zdsapi.WorkloadRequest_Del{
+			Del: &zdsapi.DelWorkload{
+				Uid: uid,
+			},
+		},
+	}
+	data, err := proto.Marshal(r)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("sending delete pod to ztunnel: %s %v", uid, r)
+
+	var delErr []error
+
+	z.conns.mu.Lock()
+	defer z.conns.mu.Unlock()
+	for conn := range z.conns.connectionSet {
+		_, err := conn.send(ctx, data, nil)
+		if err != nil {
+			delErr = append(delErr, err)
+		}
+	}
+	return errors.Join(delErr...)
+}
+
+func (z *ztunnelServer) PodAdded(ctx context.Context, uid string, netns Netns) error {
+	latestConn := z.conns.LatestConn()
+	if latestConn == nil {
+		return fmt.Errorf("no ztunnel connection")
+	}
+	r := &zdsapi.WorkloadRequest{
+		Payload: &zdsapi.WorkloadRequest_Add{
+			Add: &zdsapi.AddWorkload{
+				Uid: uid,
+			},
+		},
+	}
+	log.Debugf("About to send added pod: %s to ztunnel: %v", uid, r)
+	data, err := proto.Marshal(r)
+	if err != nil {
+		return err
+	}
+
+	fd := int(netns.Fd())
+	resp, err := latestConn.send(ctx, data, &fd)
+	if err != nil {
+		return err
+	}
+
+	if resp.GetAck().GetError() != "" {
+		log.Errorf("add-workload: got ack error: %s", resp.GetAck().GetError())
+		return fmt.Errorf("got ack error: %s", resp.GetAck().GetError())
+	}
+	return nil
+}
+
+// TODO ctx is unused here
+// nolint: unparam
+func (z *ztunnelServer) sendSnapshot(ctx context.Context, conn *ZtunnelConnection) error {
+	err := z.pods.ReadCurrentPodSnapshot(func(m map[string]Netns) error {
+		for uid, netns := range m {
+			var resp *zdsapi.WorkloadResponse
+			var err error
+			if netns != nil {
+
+				fd := int(netns.Fd())
+				log.Debugf("Sending local pod %s ztunnel", uid)
+				resp, err = conn.sendMsgAndWaitForAck(&zdsapi.WorkloadRequest{
+					Payload: &zdsapi.WorkloadRequest_Add{
+						Add: &zdsapi.AddWorkload{
+							Uid: uid,
+						},
+					},
+				}, &fd)
+			} else {
+				log.Infof("netns not available for local pod %s. sending keep to ztunnel", uid)
+				resp, err = conn.sendMsgAndWaitForAck(&zdsapi.WorkloadRequest{
+					Payload: &zdsapi.WorkloadRequest_Keep{
+						Keep: &zdsapi.KeepWorkload{
```

Comment:
Normally when ztunnel is sent the initial snapshot, it will remove any outstanding pod UIDs from its internal cache that are not in the snapshot it got from the CNI agent.

`WorkloadKeep` ensures that ztunnel does not evict that workload after getting snapshot.

I don't remember the exact scenario where the netns is not available to the CNI but ztunnel should keep the workload, @yuval-k ran into it IIRC.

Reply:
it is us being defensive; the way that we restore the pod->ns cache after restart is heuristic, so this is here to make sure that if we are not able to get all the pod netns' on start-up, the ztunnel will not drop the pod if it has it.